### PR TITLE
Make Flowvar compatible with Async

### PR DESCRIPTION
**Pull request by rayman22201**
Added some docs as well 😄 
May still need some cleanup, and it needs tests, but it basically works, as seen on the stream 👍

### CODE REVIEW

File: lib/pure/concurrency/threadpool.nim [hunk 1]
```diff
@@ -206,14 +239,16 @@ proc finished(fv: FlowVarBase) =
   # the worker thread waits for "data" to be set to nil before shutting down
   owner.data = nil
 
-proc fvFinalizer[T](fv: FlowVar[T]) = finished(fv)
+proc fvFinalizer[T](fv: FlowVar[T]) = finito(fv)
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: image

File: lib/pure/concurrency/threadpool.nim [hunk 2]
```diff
@@ -262,7 +298,12 @@ proc `^`*[T](fv: FlowVar[T]): T =
     deepCopy result, src
   else:
     result = fv.blob
-  finished(fv)
+  finito(fv)
+
+proc read*[T](future: FlowVar[T]): T =
+  ## Implement just enough of the Future api to make FlowVar compatible with the async macro.
+  ## Returns the value of the future.
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: This is an implementation detail. The docs should focus on what this proc is, not how it was implemented. You can add a note about it at the bottom, but it shouldn't be the highlight.

File: lib/pure/concurrency/threadpool.nim [hunk 3]
```diff
@@ -7,7 +7,38 @@
...
+## You most likely want the ``spawn`` procedure from this module.
+## ``spawn`` creates a new task to be executed on another thread.
+## You may synchronously block while the spawned task completes
+## using the blocking apis described in this file, and then retrieve
+## the result using the `^` operator.
+## 
+## Example:
+##
+##   .. code-block::nim
+##      proc my_thread_task(s: string) : string =
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: Nitpick: camelCase (+ below as well)

File: lib/pure/concurrency/threadpool.nim [hunk 4]
```diff
@@ -7,7 +7,38 @@
...
+##        return "hello" + s 
+##
+##      let a = spawn my_thread_task("world")
+##      sync()
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: This is unnecessary AFAIK, you just need to use the ``^`` operator.

File: lib/pure/concurrency/threadpool.nim [hunk 5]
```diff
@@ -100,20 +131,22 @@ type
...
-  FlowVarBaseObj = object of RootObj
+  FlowVarBaseObj = object of FutureBase
     ready, usesSemaphore, awaited: bool
     cv: Semaphore  # for 'blockUntilAny' support
     ai: ptr AwaitInfo
     idx: int
     data: pointer  # we incRef and unref it to keep it alive; note this MUST NOT
                    # be RootRef here otherwise the wrong GC keeps track of it!
     owner: pointer # ptr Worker
+    event*: AsyncEvent # makes Flowvar work with async
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: I think this can be private :)

File: lib/pure/concurrency/threadpool.nim [hunk 6]
```diff
@@ -224,6 +259,7 @@ proc nimFlowVarSignal(fv: FlowVarBase) {.compilerProc.} =
     signal(fv.ai.cv.c)
   if fv.usesSemaphore:
     signal(fv.cv)
+  fv.event.trigger()
 
 proc awaitAndThen*[T](fv: FlowVar[T]; action: proc (x: T) {.closure.}) =
```
Comment (dom96) [inline lib/pure/concurrency/threadpool.nim]: We need to deprecate these. If you have some time please create a seperate PR to do that :)

File: lib/pure/asyncfutures.nim [hunk 7]
```diff
@@ -196,6 +196,15 @@ proc add(callbacks: var CallbackList, function: CallbackFunc) =
         last = last.next
       last.next = newCallback
 
+proc complete*(future: FutureBase) =
+  ## Completes a base ``future``.
+  #assert(not future.finished, "Future already finished, cannot finish twice.")
+  #checkFinished(future)
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: We need this, at the very least the above assert will do.

File: lib/pure/asyncfutures.nim [hunk 8]
```diff
@@ -196,6 +196,15 @@ proc add(callbacks: var CallbackList, function: CallbackFunc) =
         last = last.next
       last.next = newCallback
 
+proc complete*(future: FutureBase) =
+  ## Completes a base ``future``.
```
Comment (dom96) [inline lib/pure/asyncfutures.nim]: Emphasise in the docs that this likely shouldn't be used unless you really know what you're doing :)

**Comment by narimiran** [comment]
Any update on this one?

**Comment by rayman22201** [comment]
@narimiran We got a bit demotivated because of the Chronos vs. Stdlib async drama.
This PR is moot because two things need to happen: (Not in any particular order)

1. the stdlib threadpool needs to be replaced with a better implementation (probably @yglukhov's https://github.com/yglukhov/threadpools)

2. The stdlib ioselectors needs to be fixed so that it does not eat up all of the socket fd's on your system. As described by @Zevv in his comment above: https://github.com/nim-lang/Nim/pull/11724#commitcomment-34295377

**Comment by dom96** [comment]
@rayman22201 is #1 necessary to get this working too?

**Comment by rayman22201** [comment]
@dom96. Not at all. I didn't mean to imply that they have to be done in some order. Both tasks are independent. 

Doing #1 will probably require a rework of this PR though. I personally would rather close this PR, do #1 first, then redo this PR based on Yuri's threadpool.

**Comment by narimiran** [comment]
@rayman22201 could you please rebase this against the latest devel?

**Comment by rayman22201** [comment]
> 
> 
> @rayman22201 could you please rebase this onto the latest devel?

@narimiran I think this particular PR is dead because of PR #12372
I will need to re-implement once PR #12372 is merged.

**Comment by KhazAkar** [comment]
> For those that missed the live stream: https://www.twitch.tv/videos/451888943
image

**Pull request closed without merging**